src/Cooked/MockChain/Balancing.hs

```diff
@@ -1,409 +1,326 @@
+    CollateralUtxosFromBalancingWallet -> case balancingWallet of
+      Nothing -> fail "Can't select collateral utxos from a balancing wallet because it does not exist."
+      Just bWallet -> (,bWallet) . Set.fromList . map fst <$> runUtxoSearch (onlyValueOutputsAtSearch bWallet)
+    CollateralUtxosFromWallet cWallet -> (,cWallet) . Set.fromList . map fst <$> runUtxoSearch (onlyValueOutputsAtSearch cWallet)
+    CollateralUtxosFromSet utxos rWallet -> return (utxos, rWallet)
 
-  -- We start with a high startingFee, but theres a chance that 'w' doesn't have
-  -- enough funds so we'll see an unbalanceable error; in that case, we switch
-  -- to the minimum fee and try again.  That feels very much like a hack, and it
-  -- is. Maybe we should witch to starting with a small fee and then increasing,
-  -- but that might require more iterations until its settled.  For now, let's
-  -- keep it just like the folks from plutus-apps did it.
-  calcFee 5 (Fee 3_000_000) skel collateralUtxosPl
-    `catchError` \case
-      -- Impossible to balance the transaction
-      MCEUnbalanceable _ _ ->
-        -- WARN "Api.minFee" takes an actual Tx but we no longer provide it
-        -- since we work on "TxSkel". However, for now, the implementation of
-        -- "Api.minFee" is a constant of 10 lovelace.
-        -- https://github.com/input-output-hk/plutus-apps/blob/d4255f05477fd8477ee9673e850ebb9ebb8c9657/plutus-ledger/src/Ledger/Index.hs#L116
-        -- forall tx. Api.minFee tx = 10 lovelace
-        calcFee 5 (Fee 10) skel collateralUtxosPl
-      -- Impossible to generate the Cardano transaction at all
-      e -> throwError e
-  where
-    -- Inspired by https://github.com/input-output-hk/plutus-apps/blob/d4255f05477fd8477ee9673e850ebb9ebb8c9657/plutus-contract/src/Wallet/Emulator/Wallet.hs#L329
+  -- At this point, the presence (or absence) of ballancing wallet dictates
+  -- whether the transaction should be automatically balanced or not.
+  (txSkelBal, fee, adjustedCollateralIns) <- case balancingWallet of
+    Nothing ->
+      -- The balancing should not be performed. We still adjust the collaterals
+      -- though around a provided fee, or the maximum fee.
+      let fee = case txOptFeePolicy txSkelOpts of
+            AutoFeeComputation -> maxFee
+            ManualFee fee' -> fee'
+       in (skelUnbal,fee,) <$> collateralInsFromFees fee collateralIns returnCollateralWallet
+    Just bWallet -> do
+      -- The balancing should be performed. We collect the balancing utxos and
+      -- filter out those already used in the unbalanced skeleton.
+      (filter ((`notElem` txSkelKnownTxOutRefs skelUnbal) . fst) -> balancingUtxos) <-
+        runUtxoSearch $ case txOptBalancingUtxos txSkelOpts of
+          BalancingUtxosAutomatic -> onlyValueOutputsAtSearch bWallet `filterWithAlways` outputTxOut
+          BalancingUtxosWith utxos -> txOutByRefSearch (Set.toList utxos) `filterWithPure` isPKOutput `filterWithAlways` outputTxOut
```

Comment:
I know this is filtered out silently but should we issue some form of warning when users declare balancing utxos that they already spend in the transaction?

Reply:
I don't think because we don't really have a way of issuing warnings in cooked (that I know of). I will add a specific comment in the skeleton and leave this open for discussion.

Reply:
I think we should consider a proper logging system in cooked to report warnings, infos, and use also for debugging. I noticed several cases where warnings could be nice to have.